### DISCUSSION TITLE
絵文字リアクションでNoteテーブルへのupdateを減らせないか、と他の更新処理について

### DISCUSSION THREAD

**Discussion by yszkst** [misskey-dev/misskey#10173] [General]
Issueにするほど確信と方針が無いのでDisucussionへ置きます。  

## Noteへのudpate  

PostgreSQLの行のUpdateは既存行に削除フラグを付けて新規に行を追加する追記型なので更新性能が低い。  
人が多いと投稿とリアクションが多いのでnoteテーブルのreactionsカラムのupdateによる負荷が多そう。  
ということで減らす方法があると良いと思っています。  

雑な思いつきとして

+ 更新が多いのは投稿されてすぐなので投稿が作成されて一定時間はRedisで更新する
+ リアクション数の整合性は捨てる   

くらいしか思いつかないので "方針が無い" です。

## 更新処理 (どこかであまりトランザクションを張ってないという投稿を見たので。  )  

参照との兼ね合いもありますが複数個所の作成更新がひとまとまりの処理はトランザクションを張ったほうがWAL掃き出し回数が減るのでやった方がよいと思います。  

計測してないので実環境だとそこまで影響はないといった情報も欲しいです。  
よろしくお願い致します。

**Comment by pumpkin2048b** [comment]
（似たようなIssue立てようか悩んでました）
リアクション関連で個人的にぼんやりと考えていたのは、
1. 更新内容を一度キューに貯める　※表示側の画面上は即時反映させる
1. 一定間隔（0.5秒とか）でキューを刈り取る
1. bulk updateに乗せて更新する

という方法を考えていました。
これで通信量と負荷は多少なりとも減るかなと。
とはいえまだmisskeyを読み解くのに精一杯で実証等はできていないですが…。
